File: backend/app/schemas/supplier.py
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, ConfigDict, EmailStr, Field, field_validator


class SupplierBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=150, description="Supplier Name")
    phone: Optional[str] = Field(None, max_length=30, description="Contact Phone Number")
    address: Optional[str] = Field(None, description="Physical Address")
    status: str = Field("active", description="Supplier status: active, inactive")
    notes: Optional[str] = None


class SupplierCreate(SupplierBase):
    supplier_code: Optional[str] = Field(None, description="Optional custom supplier code")
    opening_balance: float = Field(0.0, ge=0.0, description="Opening Due Amount ($)")


class SupplierUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=150)
    phone: Optional[str] = Field(None, max_length=30)
    address: Optional[str] = None
    current_balance: Optional[float] = Field(None, ge=0.0, description="Updated Due Balance ($)")
    status: Optional[str] = None
    notes: Optional[str] = None


class SupplierStatusUpdate(BaseModel):
    status: str = Field(..., description="Target status: active, inactive")


class SupplierResponse(SupplierBase):
    model_config = ConfigDict(from_attributes=True)

    id: str
    supplier_code: str
    company_name: Optional[str] = None
    email: Optional[str] = None
    nid: Optional[str] = None
    opening_balance: float
    current_balance: float
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/supplier.ts
export interface SupplierItem {
  id: string;
  supplier_code: string;
  name: string;
  company_name?: string;
  phone?: string;
  email?: string;
  address?: string;
  nid?: string;
  opening_balance: number;
  current_balance: number;
  status: "active" | "inactive";
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface SupplierCreatePayload {
  supplier_code?: string;
  name: string;
  phone?: string;
  address?: string;
  opening_balance?: number;
  status?: string;
  notes?: string;
}

export interface SupplierUpdatePayload {
  name?: string;
  phone?: string;
  address?: string;
  current_balance?: number;
  status?: string;
  notes?: string;
}

export interface SupplierLedgerTransaction {
  id: string;
  date: string;
  voucher_no: string;
  type: string;
  description: string;
  debit: number;
  credit: number;
  running_balance: number;
  reference_id?: string;
  reference_type?: "purchase" | "supplier_payment" | "product_return" | "balance_adjustment";
}

export interface SupplierLedgerSummary {
  opening_balance: number;
  total_purchases: number;
  total_payments: number;
  total_returns: number;
  manual_adjustments: number;
  current_due: number;
}

export interface SupplierLedgerResponse {
  supplier: SupplierItem;
  summary: SupplierLedgerSummary;
  transactions: SupplierLedgerTransaction[];
}

